core/src/main/java/org/elasticsearch/action/search/RemoteClusterService.java

```diff
@@ -0,0 +1,386 @@
+                    }
+                    remoteClusters.remove(entry.getKey());
+                    continue;
+                }
+
+                if (remote == null) { // this is a new cluster we have to add a new representation
+                    remote = new RemoteClusterConnection(settings, entry.getKey(), entry.getValue(), transportService, numRemoteConnections,
+                        nodePredicate);
+                    remoteClusters.put(entry.getKey(), remote);
+                }
+
+                // now update the seed nodes no matter if it's new or already existing
+                RemoteClusterConnection finalRemote = remote;
+                remote.updateSeedNodes(entry.getValue(), ActionListener.wrap(
+                    response -> {
+                        if (countDown.countDown()) {
+                            connectionListener.onResponse(response);
+                        }
+                    },
+                    exception -> {
+                        if (countDown.fastForward()) {
+                            connectionListener.onFailure(exception);
+                        }
+                        if (finalRemote.isClosed() == false) {
+                            logger.warn("failed to update seed list for cluster: " + entry.getKey(), exception);
+                        }
+                    }));
+            }
+        }
+        this.remoteClusters = Collections.unmodifiableMap(remoteClusters);
+    }
+
+    /**
+     * Returns <code>true</code> if at least one remote cluster is configured
+     */
+    boolean isCrossClusterSearchEnabled() {
+        return remoteClusters.isEmpty() == false;
+    }
+
+    /**
+     * Filters out indices that refer to a remote cluster and adds them to the given per cluster indices map.
+     *
+     * @param perClusterIndices a map to fill with remote cluster indices from the given request indices
+     * @param requestIndices the indices in the search request to filter
+     * @return all indices in the requestIndices array that are not remote cluster indices
+     */
+    public String[] filterIndices(Map<String, List<String>> perClusterIndices, String[] requestIndices) {
+        List<String> localIndicesList = new ArrayList<>();
+        for (String index : requestIndices) {
+            int i = index.indexOf(REMOTE_CLUSTER_INDEX_SEPARATOR);
+            if (i >= 0) {
+                String remoteCluster = index.substring(0, i);
+                if (isRemoteClusterRegistered(remoteCluster)) {
+                    String remoteIndex = index.substring(i + 1);
+                    List<String> indices = perClusterIndices.get(remoteCluster);
+                    if (indices == null) {
+                        indices = new ArrayList<>();
+                        perClusterIndices.put(remoteCluster, indices);
+                    }
+                    indices.add(remoteIndex);
+                } else {
+                    localIndicesList.add(index);
```

Comment:
+ 1 on commenting, I had it strict in the beginning but that made it impossible to ever access local indices with a `:` in their name (which unfortunately still may exist, same for aliases). So now at least we stay local whenever no matching remote cluster is configured.

Considering that configured remote cluster names are aliases and not the actual cluster names, I think it is easy to disambiguate these problems if they do happen. we could introduce some reserved `local` cluster alias that always goes local, but I am not sure it is worth it at this stage.